go/parquet/metadata/file.go

```diff
@@ -0,0 +1,478 @@
+				orders = append(orders, parquet.ColumnOrders.Undefined)
+			}
+		}
+	} else {
+		orders = orders[:f.Schema.NumColumns()]
+		orders[0] = parquet.ColumnOrders.Undefined
+		for i := 1; i < len(orders); i *= 2 {
+			copy(orders[i:], orders[:i])
+		}
+	}
+	f.Schema.UpdateColumnOrders(orders)
+}
+
+// WriterVersion returns the constructed application version from the
+// created by string
+func (f *FileMetaData) WriterVersion() *AppVersion {
+	if f.version == nil {
+		f.version = NewAppVersion(f.GetCreatedBy())
+	}
+	return f.version
+}
+
+// SetFilePath will set the file path into all of the columns in each row group.
+func (f *FileMetaData) SetFilePath(path string) {
+	for _, rg := range f.RowGroups {
+		for _, chunk := range rg.Columns {
+			chunk.FilePath = &path
+		}
+	}
+}
+
+// AppendRowGroups will add all of the rowgroup metadata from other to the
+// current file metadata
+func (f *FileMetaData) AppendRowGroups(other *FileMetaData) error {
+	if !f.Schema.Equals(other.Schema) {
+		return xerrors.New("parquet/FileMetaData: AppendRowGroups requires equal schemas")
+	}
+
+	f.RowGroups = append(f.RowGroups, other.GetRowGroups()...)
+	for _, rg := range other.GetRowGroups() {
+		f.NumRows += rg.NumRows
+	}
+	return nil
+}
+
+// Subset will construct a new FileMetaData object containing only the requested
+// row groups by index
+func (f *FileMetaData) Subset(rowGroups []int) (*FileMetaData, error) {
+	for _, i := range rowGroups {
+		if i < len(f.RowGroups) {
+			continue
+		}
+		return nil, xerrors.Errorf("parquet: this file only has %d row groups, but requested a subset including row group: %d", len(f.RowGroups), i)
+	}
+
+	out := &FileMetaData{
+		&format.FileMetaData{
+			Schema:                   f.FileMetaData.Schema,
+			CreatedBy:                f.CreatedBy,
+			ColumnOrders:             f.GetColumnOrders(),
+			EncryptionAlgorithm:      f.FileMetaData.EncryptionAlgorithm,
+			FooterSigningKeyMetadata: f.FooterSigningKeyMetadata,
+			Version:                  f.Version,
+			KeyValueMetadata:         f.KeyValueMetadata(),
+		},
+		f.version,
+		f.Schema,
+		f.FileDecryptor,
+		0,
+	}
+
+	out.RowGroups = make([]*format.RowGroup, 0, len(rowGroups))
+	for _, selected := range rowGroups {
+		out.RowGroups = append(out.RowGroups, f.RowGroups[selected])
+		out.NumRows += f.RowGroups[selected].GetNumRows()
+	}
+
+	return out, nil
+}
+
+func (f *FileMetaData) Equals(other *FileMetaData) bool {
+	return reflect.DeepEqual(f.FileMetaData, other.FileMetaData)
+}
+
+func (f *FileMetaData) KeyValueMetadata() KeyValueMetadata {
+	return f.GetKeyValueMetadata()
+}
+
+// VerifySignature constructs a cryptographic signature using the FileDecryptor
+// of the footer and then verifies it's integrity.
+//
+// Panics if f.FileDecryptor is nil
+func (f *FileMetaData) VerifySignature(signature []byte) bool {
+	if f.FileDecryptor == nil {
+		panic("decryption not set propertly, cannot verify signature")
+	}
+
+	serializer := thrift.NewThriftSerializer()
```

Comment:
@ggershinsky not sure if you want to look at cryptographic related functions here at all.

Reply:
A massive one, but overall looks well structured and in line with the Java and C++ implementations. The only way to verify the crypto code is to try to read encrypted files produced by the Java implementation, and to make the latter read encrypted files produced by this implementation. 
(C++ version might be also ok for this purpose, but the Java one is more actively maintained).
Thanks @zeroshade and @emkornfield .

Reply:
@ggershinsky I used the files here: https://github.com/apache/parquet-testing to test the encryption stuff the same way that I believe the C++ does its testing of the encrypted reading/writing. After this gets merged and I add the file module which has all of the direct file handling, I can add back the tests which did this verification.

Reply:
yep, makes sense. these files were written by C++ encryption code, but were tested against the Java version too, so should be fine.